core/src/main/java/bisq/core/api/CoreTradesService.java

```diff
@@ -116,26 +137,109 @@ void confirmPaymentReceived(String tradeId) {
         }
     }
 
-    @SuppressWarnings("unused")
     void keepFunds(String tradeId) {
-        log.info("TODO");
+        verifyTradeIsNotClosed(tradeId);
+        var trade = getOpenTrade(tradeId).orElseThrow(() ->
+                new IllegalArgumentException(format("trade with id '%s' not found", tradeId)));
+        log.info("Keeping funds received from trade {}", tradeId);
+        tradeManager.onTradeCompleted(trade);
     }
 
-    @SuppressWarnings("unused")
-    void withdrawFunds(String tradeId, String address) {
-        log.info("TODO");
+    void withdrawFunds(String tradeId, String toAddress) {
+        // An encrypted wallet must be unlocked for this operation.
+        verifyTradeIsNotClosed(tradeId);
+        var trade = getOpenTrade(tradeId).orElseThrow(() ->
+                new IllegalArgumentException(format("trade with id '%s' not found", tradeId)));
+
+        verifyIsValidBTCAddress(toAddress);
+
+        var fromAddressEntry = btcWalletService.getOrCreateAddressEntry(trade.getId(), TRADE_PAYOUT);
+        verifyFundsNotWithdrawn(fromAddressEntry);
+
+        var amount = trade.getPayoutAmount();
+        var fee = getEstimatedTxFee(fromAddressEntry.getAddressString(), toAddress, amount);
+        var receiverAmount = amount.subtract(fee);
+
+        log.info(format("Withdrawing funds received from trade %s:"
+                        + "%n From %s%n To %s%n Amt %s%n Tx Fee %s%n Receiver Amt %s",
+                tradeId,
+                fromAddressEntry.getAddressString(),
+                toAddress,
+                amount.toFriendlyString(),
+                fee.toFriendlyString(),
+                receiverAmount.toFriendlyString()));
+
+        tradeManager.onWithdrawRequest(
+                toAddress,
+                amount,
+                fee,
+                coreWalletsService.getKey(),
+                trade,
+                () -> {
+                },
+                (errorMessage, throwable) -> {
+                    log.error(errorMessage, throwable);
+                    throw new IllegalStateException(errorMessage, throwable);
+                });
     }
 
     String getTradeRole(String tradeId) {
         return tradeUtil.getRole(getTrade(tradeId));
     }
 
     Trade getTrade(String tradeId) {
-        return tradeManager.getTradeById(tradeId).orElseThrow(() ->
-                new IllegalArgumentException(format("trade with id '%s' not found", tradeId)));
+        return getOpenTrade(tradeId).orElseGet(() ->
+                getClosedTrade(tradeId).orElseThrow(() ->
+                        new IllegalArgumentException(format("trade with id '%s' not found", tradeId))
+                ));
+    }
+
+    private Optional<Trade> getOpenTrade(String tradeId) {
+        return tradeManager.getTradeById(tradeId);
+    }
+
+    private Optional<Trade> getClosedTrade(String tradeId) {
+        return closedTradableManager.getTradableById(tradeId).map(value -> (Trade) value);
```

Comment:
There is no guarantee that the `Tradable` is a `Trade` here, probably good to add a check for `instanceof`.

Reply:
Fixed in PR https://github.com/bisq-network/bisq/pull/4731, commit https://github.com/bisq-network/bisq/pull/4731/commits/fcdfc687e45d1d613838d7189aa4e8edd41228e2.